### ISSUE TITLE
cifar training failes "No device code available for function"

### ISSUE THREAD

**Issue by witeko** [ROCmSoftwarePlatform/tensorflow-upstream#174]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: ubuntu 18.04.1
- **TensorFlow installed from (source or binary)**: rocm 1.10 whl
- **Python version**: 3.5 anaconda
- **GPU model and memory**: rx580
- **Exact command to reproduce**:
- rocm 1.9

### Describe the problem
When using the latest tensorflow_rocm 1.10 whl the cifar training failes - see below.
When using tensorflow_rocm 1.8 also whl it runs just fine.

### Source code / logs
(tf_110) witold@witold-tr:~/models-master/tutorials/image/cifar10$ python ./cifar10_train.py
Filling queue with 20000 CIFAR images before starting to train. This will take a few minutes.
2018-09-20 19:59:05.250451: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-09-20 19:59:05.255674: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1519] Found device 0 with properties: 
name: Ellesmere [Radeon RX 470/480]
AMDGPU ISA: gfx803
memoryClockRate (GHz) 1.34
pciBusID 0000:42:00.0
Total memory: 8.00GiB
Free memory: 7.75GiB
2018-09-20 19:59:05.255700: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1630] Adding visible gpu devices: 0
2018-09-20 19:59:05.255718: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1039] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-09-20 19:59:05.255725: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1045]      0 
2018-09-20 19:59:05.255729: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1058] 0:   N 
2018-09-20 19:59:05.255762: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1178] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7539 MB memory) -> physical GPU (device: 0, name: Ellesmere [Radeon RX 470/480], pci bus id: 0000:42:00.0)
terminate called recursively
terminate called after throwing an instance of 'std::runtime_error'
  what():  No device code available for function: _ZN7rocprim6detail19block_reduce_kernelILb0ENS0_21default_reduce_configILj0EfEEfNS_18transform_iteratorIPfN10tensorflow10squareHalfIfEEfEES5_fN6hipcub3SumEEEvT2_mT3_T4_T5_
Aborted (core dumped)

**Comment by sunway513** [comment]
@witeko , could you try TF1.10 docker images?
That's to verify if the issue is from your user bits or kernel bits configurations.
You can find the docker images here:
https://hub.docker.com/r/rocm/tensorflow/

In addition, please provide the rocminfo log on your system:
`/opt/rocm/bin/rocminfo`

**Comment by witeko** [comment]
@sunway513 , rocminfo below.
I've never used docker before, and unfortunatelly it seems that the description is not 100% full (even with: https://github.com/RadeonOpenCompute/ROCm-docker/blob/master/quick-start.md).
as I call it: "Sbd saved 2 bits of storage, and generated 2h of work for me..."
I will invest those 2h just not right now...

/opt/rocm/bin/rocminfo
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.1
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (number of timestamp)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    AMD Ryzen Threadripper 1920X 12-Core Processor
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0                                  
  Queue Min Size:          0                                  
  Queue Max Size:          0                                  
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768KB                            
  Chip ID:                 0                                  
  Cacheline Size:          64                                 
  Max Clock Frequency (MHz):0                                  
  BDFID:                   0                                  
  Compute Unit:            24                                 
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    65904380KB                         
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    65904380KB                         
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        TRUE                               
  ISA Info:                
    N/A                      
*******                  
Agent 2                  
*******                  
  Name:                    gfx803                             
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128                                
  Queue Min Size:          4096                               
  Queue Max Size:          131072                             
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      16KB                               
  Chip ID:                 26591                              
  Cacheline Size:          64                                 
  Max Clock Frequency (MHz):1340                               
  BDFID:                   16896                              
  Compute Unit:            36                                 
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      FALSE                              
  Wavefront Size:          64                                 
  Workgroup Max Size:      1024                               
  Workgroup Max Size Per Dimension:
    Dim[0]:                  67109888                           
    Dim[1]:                  1107297280                         
    Dim[2]:                  0                                  
  Grid Max Size:           4294967295                         
  Waves Per CU:            40                                 
  Max Work-item Per CU:    2560                               
  Grid Max Size per Dimension:
    Dim[0]:                  4294967295                         
    Dim[1]:                  4294967295                         
    Dim[2]:                  4294967295                         
  Max number Of fbarriers Per Workgroup:32                                 
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    8388608KB                          
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Acessible by all:        FALSE                              
    Pool 2                   
      Segment:                 GROUP                              
      Size:                    64KB                               
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Alignment:         0KB                                
      Acessible by all:        FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx803          
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Dimension: 
        Dim[0]:                  67109888                           
        Dim[1]:                  1024                               
        Dim[2]:                  16777217                           
      Workgroup Max Size:      1024                               
      Grid Max Dimension:      
        x                        4294967295                         
        y                        4294967295                         
        z                        4294967295                         
      Grid Max Size:           4294967295                         
      FBarrier Max Size:       32                                 
*** Done ***

**Comment by sunway513** [comment]
@witeko thanks for the logs.
Actually, it's pretty straightforward to set up docker, here are the three commands you can finish in 2-mins (depends on your network speed for pulling docker image):
`
# To configure docker environment
curl -sSL https://get.docker.com/ | sh
# To run TF docker image
alias drun='sudo docker run -it --network=host --device=/dev/kfd --device=/dev/dri --group-add video --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -v $HOME/dockerx:/dockerx -v /data/imagenet/tf:/imagenet'
drun rocm/tensorflow:latest
`

**Comment by witeko** [comment]
yes, Ive alredy run them.
what next? how to open my anaconda env?
as a reply: walking is also easy, but children need to actually learn it...

**Comment by sunway513** [comment]
@witeko , thanks for the heads up!
We've fixed the docker container and working on updating the whl packages with `gfx803` device code support.
The following command would update your docker container:
`sudo docker pull rocm/tensorflow:latest`

**Comment by witeko** [comment]
@sunway513 thanks for Your help, I may be a little bit harsh, but I can appreciate sbds effort :)
I assume it will work just fine once You fix everything. :)
I will check then.

**Comment by sunway513** [comment]
@witeko We have fixed the whl package hosted by pypl, the following command should fix your bare metal TF setup:
`
pip3 uninstall -y tensorflow-rocm && pip3 install --user tensorflow-rocm
`

**Comment by witeko** [comment]
@sunway513 looks good now :) ty :)

You now have an additional "warning" during tf installation:
keras 2.2.2 has requirement keras-applications==1.0.4, but you'll have keras-applications 1.0.5 which is incompatible.
keras 2.2.2 has requirement keras-preprocessing==1.0.2, but you'll have keras-preprocessing 1.0.3 which is incompatible.

I think it's from tensorflow itself (not related to rocm).


root@witold-tr:/root/models/tutorials/image/cifar10# python3 cifar10_train.py 
Filling queue with 20000 CIFAR images before starting to train. This will take a few minutes.
2018-09-20 22:37:32.836168: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1519] Found device 0 with properties: 
name: Device 67df
AMDGPU ISA: gfx803
memoryClockRate (GHz) 1.34
pciBusID 0000:42:00.0
Total memory: 8.00GiB
Free memory: 7.75GiB
2018-09-20 22:37:32.836208: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1630] Adding visible gpu devices: 0
2018-09-20 22:37:32.836224: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1039] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-09-20 22:37:32.836231: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1045]      0 
2018-09-20 22:37:32.836236: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1058] 0:   N 
2018-09-20 22:37:32.836273: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1178] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7539 MB memory) -> physical GPU (device: 0, name: Device 67df, pci bus id: 0000:42:00.0)
2018-09-20 22:37:42.084948: I tensorflow/core/kernels/conv_grad_filter_ops.cc:974] running auto-tune for Backward-Filter
2018-09-20 22:37:43.478271: I tensorflow/core/kernels/conv_grad_input_ops.cc:1021] running auto-tune for Backward-Data
2018-09-20 22:37:45.318486: I tensorflow/core/kernels/conv_grad_filter_ops.cc:974] running auto-tune for Backward-Filter
2018-09-20 22:37:46.774014: step 0, loss = 4.68 (91.3 examples/sec; 1.402 sec/batch)
2018-09-20 22:37:47.020210: step 10, loss = 4.61 (5199.1 examples/sec; 0.025 sec/batch)
2018-09-20 22:37:47.177011: step 20, loss = 4.47 (8163.2 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:47.334010: step 30, loss = 4.38 (8152.9 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:47.491681: step 40, loss = 4.38 (8118.2 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:47.648637: step 50, loss = 4.28 (8155.2 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:47.806500: step 60, loss = 4.24 (8108.3 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:47.963793: step 70, loss = 4.58 (8137.7 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:48.121788: step 80, loss = 4.16 (8101.5 examples/sec; 0.016 sec/batch)
2018-09-20 22:37:48.278778: step 90, loss = 4.19 (8153.4 examples/sec; 0.016 sec/batch)

**Comment by witeko** [comment]
also in my anaconda works

**Comment by sunway513** [comment]
@witeko Very cool :-) 
I'll close the issue then, thanks!
